modules/core/src/main/scala/com.snowplowanalytics.iglu/client/validator/CirceValidator.scala

```diff
@@ -102,4 +106,112 @@ object CirceValidator extends Validator[Json] {
 
   private def fromValidationMessage(m: ValidationMessage): ValidatorReport =
     ValidatorReport(m.getMessage, m.getPath.some, m.getArguments.toList, m.getType.some)
+
+  private def evaluateSchema(
+    schemaAsNode: JsonNode
+  ): Either[ValidatorError.InvalidSchema, JsonSchema] = {
+    Either
+      .catchNonFatal(
+        IgluMetaschemaFactory
+          .getSchema(schemaAsNode, SchemaValidatorsConfig)
+      )
+      .leftMap(ValidatorError.schemaIssue)
+  }
+
+  private def validateSchemaAgainstV4(schema: JsonNode): List[ValidatorError.SchemaIssue] = {
+    V4Schema
+      .validate(schema)
+      .asScala
+      .toList
+      .map(m => ValidatorError.SchemaIssue(m.getPath, m.getMessage))
+  }
+
+  private[client] object WithCaching {
+    type SchemaEvaluationKey         = (SchemaKey, TTL)
+    type SchemaEvaluationResult      = Either[ValidatorError.InvalidSchema, JsonSchema]
+    type SchemaEvaluationCache[F[_]] = LruMap[F, SchemaEvaluationKey, SchemaEvaluationResult]
+    type InitValidatorCache[F[_]]    = CreateLruMap[F, SchemaEvaluationKey, SchemaEvaluationResult]
+
+    def validate[F[_]: Monad](
+      schemaEvaluationCache: SchemaEvaluationCache[F]
+    )(data: Json, schema: SchemaLookupResult): F[Either[ValidatorError, Unit]] = {
+      getFromCacheOrEvaluate(schemaEvaluationCache)(schema)
+        .map {
+          _.flatMap { jsonschema =>
+            validateOnReadySchema(jsonschema, data)
+              .leftMap(ValidatorError.InvalidData.apply)
+          }
+        }
+    }
+
+    private def getFromCacheOrEvaluate[F[_]: Monad](
+      evaluationCache: SchemaEvaluationCache[F]
+    )(result: SchemaLookupResult): F[Either[ValidatorError.InvalidSchema, JsonSchema]] = {
+      result match {
+        case ResolverResult.Cached(key, schema, timestamp) =>
+          evaluationCache.get((key, timestamp)).flatMap {
+            case Some(alreadyEvaluatedSchema) =>
+              alreadyEvaluatedSchema.pure[F]
+            case None =>
+              provideNewJsonSchema(schema)
+                .pure[F]
+                .flatTap(result => evaluationCache.put((key, timestamp), result))
+          }
+        case ResolverResult.NotCached(schema) =>
+          provideNewJsonSchema(schema).pure[F]
+      }
+    }
+    def validate2[F[_]: Monad](
```

Comment:
This one is unused?

Reply:
Ooooops 😳 . Thanks for spotting this; I've removed it now.